flang/test/Lower/array-expression.f90

```diff
@@ -452,4 +452,107 @@ subroutine test_assigning_to_assumed_shape_slices(x)
+  ! CHECK: fir.array_merge_store %{{.*}}, %{{.*}} to %{{.*}} typeparams %{{.*}} : !fir.ref<!fir.array<40x!fir.char<1,?>>>, i32
+
+  a = b
+end subroutine test19d
+
+! CHECK-LABEL: func @_QPtest19e
+subroutine test19e(a,b)
+  character(*) a(50)
+  character(*) b(50)
+
+  ! CHECK-DAG: fir.array_load %{{.*}}(%{{.*}}) typeparams %{{.*}}#1 : (!fir.ref<!fir.array<50x!fir.char<1,?>>>, !fir.shape<1>, index) -> !fir.array<50x!fir.char<1,?>>
+  ! CHECK-DAG: fir.array_load %{{.*}}(%{{.*}}) typeparams %{{.*}}#1 : (!fir.ref<!fir.array<50x!fir.char<1,?>>>, !fir.shape<1>, index) -> !fir.array<50x!fir.char<1,?>>
+  ! CHECK: fir.do_loop %{{.*}} = %{{.*}} to %{{.*}} step %{{.*}} iter_args(%{{.*}} = %{{.*}}) -> (!fir.array<50x!fir.char<1,?>>) {
+  ! CHECK: fir.array_fetch %{{.*}}, %{{.*}} typeparams %{{.*}}#1 : (!fir.array<50x!fir.char<1,?>>, index, index) -> !fir.ref<!fir.char<1,?>>
+  ! CHECK: fir.array_update %{{.*}}, %{{.*}}, %{{.*}} typeparams %{{.*}}#1 : (!fir.array<50x!fir.char<1,?>>, !fir.ref<!fir.char<1,?>>, index, index) -> !fir.array<50x!fir.char<1,?>>
+  ! CHECK: fir.array_merge_store %{{.*}}, %{{.*}} to %{{.*}} typeparams %{{.*}}#1 : !fir.ref<!fir.array<50x!fir.char<1,?>>>, index
+
+  a = b
+end subroutine test19e
+
+! CHECK-LABEL: func @_QPtest19f
+subroutine test19f(a,b)
+  character(*) a(60)
+  character(*) b(60)
+
+  ! CHECK-DAG: fir.array_load %{{.*}}(%{{.*}}) typeparams %{{.*}} : (!fir.ref<!fir.array<60x!fir.char<1,?>>>, !fir.shape<1>, index) -> !fir.array<60x!fir.char<1,?>>
+  ! CHECK-DAG: fir.address_of(@_QQcl.70726566697820) : !fir.ref<!fir.char<1,7>>
+  ! CHECk-DAG: fir.array_load %{{.*}}(%{{.*}}) typeparams %{{.*}} : (!fir.ref<!fir.array<60x!fir.char<1,?>>>, !fir.shape<1>, index) -> !fir.array<60x!fir.char<1,?>>
+  ! CHECK: fir.do_loop %{{.*}} = %{{.*}} to %{{.*}} step %{{.*}} iter_args(%{{.*}} = %{{.*}}) -> (!fir.array<60x!fir.char<1,?>>) {
+  ! CHECK: fir.array_fetch %{{.*}}, %{{.*}} typeparams %{{.*}} : (!fir.array<60x!fir.char<1,?>>, index, index) -> !fir.ref<!fir.char<1,?>>
+  ! CHECK: fir.call @llvm.memmove.p0i8.p0i8.i64(%{{.*}}, %{{.*}}, %{{.*}}, %false) : (!fir.ref<i8>, !fir.ref<i8>, i64, i1) -> ()
+  ! CHECK: fir.do_loop %{{.*}} = %{{.*}} to %{{.*}} step %{{.*}} {
+  ! CHECK: fir.load %{{.*}} : !fir.ref<!fir.char<1>>
+  ! CHECK: fir.store %{{.*}} to %{{.*}} : !fir.ref<!fir.char<1>>
+  ! CHECK: fir.array_update %{{.*}}, %{{.*}}, %{{.*}} typeparams %{{.*}} : (!fir.array<60x!fir.char<1,?>>, !fir.ref<!fir.char<1,?>>, index, index) -> !fir.array<60x!fir.char<1,?>>
+  ! CHECK: fir.array_merge_store %{{.*}}, %{{.*}} to %{{.*}} typeparams %{{.*}} : !fir.ref<!fir.array<60x!fir.char<1,?>>>, index
+  a = "prefix " // b
+end subroutine test19f
+
+! CHECK-LABEL: func @_QPtest19g
+subroutine test19g(a,b,i)
+  character(kind=4,len=i) a(70)
+  character(kind=2,len=13) b(140)
+  a = b(1:140:2)
+end subroutine test19g
```

Comment:
I think it would be worth trying some cases with assumed shape character arrays (involves fir.box and discontinuity). Both constant and dynamic length assumed shape could have fun effects :).

Reply:
Hmm. I guess I didn't copy that over from my small working tests.